### PR TITLE
Add promoter expression

### PR DESCRIPTION
**Pull request by SteveBronder**
@WardBrian I got this up to the point that I think I have all the backend setup. What I'm having trouble figuring out is where to place the actual promotions in the MIR lol! 

The thing I'm confused by is where to get the udf signature information. tmk I would need a separate pass where I also include the functions block as I go over each of the other blocks. I was at first thinking about doing this in `Stan_math_backend/Transform_mir` where I'd have some functions like the below that do go over each block/statement/function/expression adding promotions where we need them.

```ocaml

(*Go over each block updating expressions in functions in statements that need promoted *)
let promote_func_args fun_block prog_block = 
  (*make a table for each udf and it's signature*)
  udf_func_signatures = get_udf_signatures fun_block in 
  (*Go over each statement updating it's functions whose args need promoted *)
  List.map ~f:(promote_func_arg_stmts udf_func_signatures) prog_block

(*iterate over each block in the program doing promotion*)
let do_program_promotion prog = 
   let promote_func_args' = promote_func_args p.functions_block
  Program.map_stmts promote_func_args' p
```

But should this happen somewhere else sooner? Like should we do it right after we call Ast_to_Mir?

Also do we only need to do this promotion for UDFs or should it also happen for stan math functions. UDFs are a little simpler as there is only ever one signature. Stan math will be a little harder as we would need to choose the "most appropriate" signature

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [x] OR, no user-facing changes were made

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by WardBrian** [comment]
@SteveBronder - if you want to revisit this soon I think it is easier to do in the typechecker now per my above comment. Happy to help if needed

**Comment by SteveBronder** [comment]
Yes! Got lost in pathfinder I'll come back to this in the coming week

**Comment by WardBrian** [comment]
I've linked #373 to this as this is essentially what it was describing, just now applied to the newer types

**Comment by WardBrian** [comment]
Looks like I confused Github pretty seriously somehow. I've got this hooked up to the typechecker in an ok-ish way at any rate

**Comment by WardBrian** [comment]
@SteveBronder getting errors on the code gen of this:

```
/integration/good/function-signatures/math/matrix/Phi.o ../../test/integration/good/function-signatures/math/matrix/Phi.hpp
../../test/integration/good/function-signatures/math/matrix/Phi.hpp:797:41: error: too few template arguments for alias template 'promote_scalar_t'
      assign(trans_x2z, Phi(stan::math::promote_scalar_t<double>(x1z)),
                                        ^
/mnt/nvme1n1/workspace/stanc3_PR-1004/performance-tests-cmdstan/cmdstan/stan/lib/stan_math/stan/math/prim/meta/promote_scalar_type.hpp:97:1: note: template is declared here
using promote_scalar_t =
^
../../test/integration/good/function-signatures/math/matrix/Phi.hpp:808:41: error: too few template arguments for alias template 'promote_scalar_t'
      assign(trans_x2w, Phi(stan::math::promote_scalar_t<double>(x1w)),
                                        ^
/mnt/nvme1n1/workspace/stanc3_PR-1004/performance-tests-cmdstan/cmdstan/stan/lib/stan_math/stan/math/prim/meta/promote_scalar_type.hpp:97:1: note: template is declared here
using promote_scalar_t =
^
```

Could you take a look?

**Comment by WardBrian** [comment]
@SteveBronder this is passing the trivial tests I set up. If you can take another pass at it there's definitely some clean up it could use.

File: src/frontend/SignatureMismatch.ml [hunk 1]
```diff
@@ -153,6 +166,16 @@ let extract_function_types f =
       Some (return, args, (fun x -> UserDefined x), mem)
   | _ -> None
 
+let promote es promotions =
+  List.map2_exn es promotions ~f:(fun (exp : Ast.typed_expression) prom ->
+      let emeta = exp.emeta in
+      match prom with
+      | RealPromotion when emeta.type_ <> UReal ->
+          Ast.{expr= Ast.Promotion (exp, UReal); emeta}
```
Comment (nhuurre) [inline src/frontend/SignatureMismatch.ml:174]: You check for `UReal` but what if `emeta.type_` is `UArray UReal`?
And shouldn't the expression type actually change, update `emeta`?

**Review by nhuurre — COMMENTED** [review]
Ternary operator `x ? y : z` and array literals `{x, y, z}` may also generate `stan::math::promote_scalar`. Would it be feasible to implement those too as a promotion in the MIR?

File: src/middle/Expr.ml [hunk 2]
```diff
@@ -17,6 +17,7 @@ module Fixed = struct
       | EAnd of 'a * 'a
       | EOr of 'a * 'a
       | Indexed of 'a * 'a Index.t list
+      | Promotion of 'a * UnsizedType.t
```
[the remark this answers is not in the document]
Comment (WardBrian) [reply src/middle/Expr.ml:20]: I'd like @SteveBronder to weigh in on this and also your questions regarding `?:` and array literals. I think the answer might be that the autodifftype is not changed from the original type, so we don't store it? 

Though you make a good point about the type information as well. I think to a certain extent it is nice from a code-readability standpoint to have it be part of the MIR node itself, but I could be swayed.

[the remark this answers is not in the document]
Comment (WardBrian) [reply src/middle/Expr.ml]: I looked into it, and I think the ternary case could require promotions to contain adlevels not from the original expression, so I think it's worth folding into the type using the signature you propose. I think I've got that now.

**Review by nhuurre — APPROVED** [review]
Looks good to me.

**Comment by WardBrian** [comment]
Thanks @nhuurre. I’m gonna add some more tests when I fix the merge and update some of the doc. I’ll merge after the tests are done and the doc is approved

**Pull request merged**